src/main/groovy/com/deploygate/gradle/plugins/tasks/LoginTask.groovy

```diff
@@ -3,33 +3,62 @@ package com.deploygate.gradle.plugins.tasks
 import com.deploygate.gradle.plugins.dsl.DeployGateExtension
 import com.deploygate.gradle.plugins.internal.http.ApiClient
 import com.deploygate.gradle.plugins.internal.http.GetCredentialsResponse
+import com.deploygate.gradle.plugins.tasks.inputs.Credentials
 import com.deploygate.gradle.plugins.utils.BrowserUtils
 import com.deploygate.gradle.plugins.utils.UrlUtils
 import com.sun.net.httpserver.HttpExchange
 import com.sun.net.httpserver.HttpServer
 import org.gradle.api.DefaultTask
 import org.gradle.api.GradleException
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
+import org.jetbrains.annotations.NotNull
+import org.jetbrains.annotations.VisibleForTesting
 
+import javax.inject.Inject
 import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 
-class LoginTask extends DefaultTask {
+abstract class LoginTask extends DefaultTask {
+
+    @Input
```

Comment:
Gradle look up properties annotated with `Input` to calculate `dirtiness` for the cache.
We can use direct type like `String` but it's not preferable for lazy configuration. Instead, Gradle introduces `Property` class for mutation and Provider class for readonly. And also, they have many annotations...

`Internal` can be used if you want to `exclude` the variable from the cache. For example, onetime flag in a task should be annotated with `Internal` because it's not an input.
`Optional` means just nullable.

For the more detail https://docs.gradle.org/current/userguide/lazy_configuration.html